File: test/TEST_IpcResponse.cs
using System;
using System.Collections;
using System.Collections.Generic;

using Xunit;

using InoIPC;

namespace InoIPC.TEST
{
   // ============================================================
   /// <summary>
   /// Tests for IpcResponse builders.
   /// </summary>
   // ============================================================
   public class TEST_IpcResponse
   {

   #region Success

      [Fact]
      public void Success_Empty()
      {
         var r = IpcResponse.Success();

         Assert.True(r.IsSuccess);
         Assert.Contains("\"success\":true", r.RawJson);
      }

      [Fact]
      public void Success_Message()
      {
         var r = IpcResponse.Success("Connected");

         Assert.True(r.IsSuccess);
         Assert.Contains("\"message\":\"Connected\"", r.RawJson);
      }

      [Fact]
      public void Success_KeyValue()
      {
         var r = IpcResponse.Success("port", 56400);

         Assert.True(r.IsSuccess);
         Assert.Contains("\"port\":56400", r.RawJson);
      }

      [Fact]
      public void Success_Dict()
      {
         var r = IpcResponse.Success(new Dictionary<string, object>
         {
            ["port"] = 56400,
            ["host"] = "localhost"
         });

         Assert.True(r.IsSuccess);
         Assert.Contains("\"port\":56400", r.RawJson);
         Assert.Contains("\"host\":\"localhost\"", r.RawJson);
      }

      [Fact]
      public void Success_KeyValue_ReservedKey_Throws()
      {
         Assert.Throws<ArgumentException>(() => IpcResponse.Success("success", true));
      }

      [Fact]
      public void Success_Dict_ReservedKey_Throws()
      {
         Assert.Throws<ArgumentException>(() => IpcResponse.Success(new Dictionary<string, object>
         {
            ["success"] = true
         }));
      }

   #endregion

   #region Error

      [Fact]
      public void Error_CodeMessage()
      {
         var r = IpcResponse.Error("TIMEOUT", "timed out");

         Assert.False(r.IsSuccess);
         Assert.Contains("\"code\":\"TIMEOUT\"", r.RawJson);
         Assert.Contains("\"message\":\"timed out\"", r.RawJson);
      }

      [Fact]
      public void Error_WithData()
      {
         var r = IpcResponse.Error("TIMEOUT", "timed out", new Dictionary<string, object>
         {
            ["elapsed"] = 5000
         });

         Assert.False(r.IsSuccess);
         Assert.Contains("\"elapsed\":5000", r.RawJson);
      }

      [Fact]
      public void Error_Data_ReservedKey_Throws()
      {
         Assert.Throws<ArgumentException>(() => IpcResponse.Error("X", "x", new Dictionary<string, object>
         {
            ["code"] = "override"
         }));
      }

   #endregion

   #region Parse

      [Fact]
      public void Parse_Success()
      {
         var r = IpcResponse.Parse("{\"success\":true,\"port\":56400}");

         Assert.True(r.IsSuccess);
      }

      [Fact]
      public void Parse_Error()
      {
         var r = IpcResponse.Parse("{\"success\":false,\"error\":{\"code\":\"X\"}}");

         Assert.False(r.IsSuccess);
      }

      [Fact]
      public void Parse_InvalidJson()
      {
         var r = IpcResponse.Parse("not json");

         Assert.False(r.IsSuccess);
      }

   #endregion

   }
}

File: rust/inoipc/tests/test_ipc_response.rs
use std::collections::HashMap;

use serde_json::json;

use inoipc::IpcResponse;

// ----------------------------------------------------------------
// Success builders
// ----------------------------------------------------------------

#[test]
fn success_empty() {
    let r = IpcResponse::success();
    assert!(r.is_success());
    assert!(r.raw_json().contains("\"success\":true"));
}

#[test]
fn success_message() {
    let r = IpcResponse::success_msg("Connected");
    assert!(r.is_success());
    assert!(r.raw_json().contains("\"message\":\"Connected\""));
}

#[test]
fn success_kv() {
    let r = IpcResponse::success_kv("port", json!(56400)).unwrap();
    assert!(r.is_success());
    assert!(r.raw_json().contains("\"port\":56400"));
}

#[test]
fn success_map() {
    let mut data = HashMap::new();
    data.insert("port".to_string(), json!(56400));
    data.insert("host".to_string(), json!("localhost"));
    let r = IpcResponse::success_map(&data).unwrap();
    assert!(r.is_success());
    assert!(r.raw_json().contains("\"port\":56400"));
    assert!(r.raw_json().contains("\"host\":\"localhost\""));
}

#[test]
fn success_kv_reserved_key_errors() {
    let result = IpcResponse::success_kv("success", json!(true));
    assert!(result.is_err());
}

#[test]
fn success_map_reserved_key_errors() {
    let mut data = HashMap::new();
    data.insert("success".to_string(), json!(true));
    let result = IpcResponse::success_map(&data);
    assert!(result.is_err());
}

// ----------------------------------------------------------------
// Error builders
// ----------------------------------------------------------------

#[test]
fn error_code_message() {
    let r = IpcResponse::error("TIMEOUT", "timed out");
    assert!(!r.is_success());
    assert!(r.raw_json().contains("\"code\":\"TIMEOUT\""));
    assert!(r.raw_json().contains("\"message\":\"timed out\""));
}

#[test]
fn error_with_data() {
    let mut data = HashMap::new();
    data.insert("elapsed".to_string(), json!(5000));
    let r = IpcResponse::error_with("TIMEOUT", "timed out", &data).unwrap();
    assert!(!r.is_success());
    assert!(r.raw_json().contains("\"elapsed\":5000"));
}

#[test]
fn error_data_reserved_key_errors() {
    let mut data = HashMap::new();
    data.insert("code".to_string(), json!("override"));
    let result = IpcResponse::error_with("X", "x", &data);
    assert!(result.is_err());
}

// ----------------------------------------------------------------
// Parse
// ----------------------------------------------------------------

#[test]
fn parse_success() {
    let r = IpcResponse::parse("{\"success\":true,\"port\":56400}");
    assert!(r.is_success());
}

#[test]
fn parse_error() {
    let r = IpcResponse::parse("{\"success\":false,\"error\":{\"code\":\"X\"}}");
    assert!(!r.is_success());
}

#[test]
fn parse_invalid_json() {
    let r = IpcResponse::parse("not json");
    assert!(!r.is_success());
}
